algorithms/tree_algorithm.py

```diff
@@ -0,0 +1,187 @@
+from supporting_components.graph import Graph, Vertex
+from math import factorial
+
+
+def is_tree(G: "Graph"):
+    """
+    Determines if graph G is a tree based on the Theorem that the amount of edges is one less than the amount of vertices
+    :return: Boolean if graph G is a tree or not
+    """
+    return len(G.edges) == len(G.vertices) - 1
+
+
+def trees_count_isomorphisms(T1: "Graph", T2: "Graph", count_flag: "Bool"):
+    """
+    :param T1, T2: Graphs that are trees (so is_tree(T1) and is_tree(T2) return True)
+    :param count_flag: Flag that determines what return argument is
+    :return: If count_flag is False, the return argument is a boolean indicating if T1 and T2 are isomorphic or not
+    If count_flag is True, the return argument is the amount of isomorphisms between T1 and T2
+    """
+    # Determine the root of both trees
+    root_T1 = __root_label(T1)
+    root_T2 = __root_label(T2)
+
+    # Tree can have one or two roots, if amount of roots is not equal, not isomorphic
+    if len(root_T1) != len(root_T2):
+        if count_flag:
+            return 0
+        return False
+
+    # If there are two roots, the amount of isomorphisms is the addition of isomorphisms from the two mappings
+    # of roots to each other. Because the maximum amount of roots is 2, all roots of T1 are mapped to one root of T2.
+    # Root levels are overwritten in the second cycle.
+    total_isomorphisms = 0
+    # for i_root in range(len(root_T1)):
+    for i_root in [0]:
+        root1 = root_T1[i_root]
+        root2 = root_T2[0]
+
+        # Assign level numbers to all nodes
+        __assign_level(root1, 0, [])
+        __assign_level(root2, 0, [])
+
+        # Assign all vertices to level number lists
+        L1 = {}
+        for v in T1:
+            L1.setdefault(v.level, []).append(v)
+        L2 = {}
+        for v in T2:
+            L2.setdefault(v.level, []).append(v)
+
+        # The number of levels of each tree should be equal, otherwise not isomorphic
+        h1, h2 = max(L1.keys()), max(L2.keys())
+        if max(L1.keys()) != max(L2.keys()):
+            continue
+        h = h1
+
+        # From the bottom level up, assign names to the vertices
+        # If the vertex is a leaf, the name 'lr' is assigned
+        # The name of all other vertices is a sorted collection of the names of its children
+        # At the same time, remember how many automorphisms the subtree of that vertex has
+        are_isomorphic = True    # For each level it is checked if the trees still can be isomoprhic, if not, the loop can be stopped
+        for i in range(h, -1, -1):
+            # Name the vertices of tree T1
+            H1 = []
+            for v in L1[i]:
+                if v.degree_fixed == 1:
```

Comment:
Is degree_fixed ook iets dat wordt gezet als je geen twin color initialization hebt?

Reply:
In de integratie test wordt dat sowieso gedaan, ik zag wel dat ik het in mijn tree_test nog niet had gedaan. Maar dat heb ik nu aangepast. Dus inderdaad ook als je geen twin removal hebt